DietPi-Software | Firefox Sync Server polish

MichaIng opened the pull request:
**Status**: Ready
- [x] DietPi-Software | Do not enable service on install as it is not DietPi-controlled
- [x] Changelog
- [x] Online docs: https://dietpi.com/phpbb/viewtopic.php?p=24713#p24713

**Commit list/description**:
+ DietPi-Services | Add support for new Firefox Sync Server
+ DietPi-Software | Firefox Sync Server: Add online docs URL: https://dietpi.com/phpbb/viewtopic.php?p=24713#p24713
+ DietPi-Software | Firefox Sync Server: Do not edit Makefile to use our config, instead call gunicorn with correct option directly
+ DietPi-Software | Firefox Sync Server: Add strict service hardening to allow R/W only to our config+database dir, the ffsync users home
+ DietPi-Software | Firefox Sync Server: Do not enable service on install, since it is now controlled by DietPi-Services
+ CHANGELOG | Firefox Sync Server has been added to our software list

MichaIng committed:
v6.31

+ DietPi-Services | Add support for new Firefox Sync Server

MichaIng committed:
v6.31

+ DietPi-Software | Firefox Sync Server: Add online docs URL: https://dietpi.com/phpbb/viewtopic.php?p=24713#p24713
+ DietPi-Software | Firefox Sync Server: Do not edit Makefile to use our config, instead call gunicorn with correct option directly
+ DietPi-Software | Firefox Sync Server: Add strict service hardening to allow R/W only to our config+database dir, the ffsync users home
+ DietPi-Software | Firefox Sync Server: Do not enable service on install, since it is now controlled by DietPi-Services

MichaIng committed:
v6.31

+ CHANGELOG | Firefox Sync Server has been added to our software list

MichaIng commented:
@CedArctic 
DietPi-Services control has been added, as well as some other enhancements, service hardening and of course changelog + online docs. Hope you like it, tell me if you have any suggestions.

And it would be awesome if you could test the final code by really syncing a Firefox with it. As said I have no Mozilla account and would skip it as long as not required 😉.

To test code of this branch:
```
dietpi-backup # Create/update your backup at best, as it's dev code
G_DEV_BRANCH ffsync # Switches branch in config and runs the update automatically
```

If any issues occur, you can revert via: `dietpi-backup 1`
However no critical changes have been done to the code so far, so should run fine.
___________
🈯 General connection test:
```
root@VM-Buster:~# curl  192.168.1.22:5000
it works!

root@VM-Buster:~# curl -I 192.168.1.22:5000
HTTP/1.1 200 OK
Server: gunicorn/19.6.0
Date: Wed, 13 May 2020 21:19:12 GMT
Connection: close
Content-Type: text/html; charset=UTF-8
Content-Length: 9
X-Weave-Timestamp: 1589404752.78
X-Timestamp: 1589404752

root@VM-Buster:~# curl  192.168.1.22:5000/token/1.0/sync/1.5
{"status": "error", "errors": [{"location": "body", "name": "", "description": "Unauthorized"}]}

root@VM-Buster:~# curl -I 192.168.1.22:5000/token/1.0/sync/1.5
HTTP/1.1 401 Unauthorized
Server: gunicorn/19.6.0
Date: Wed, 13 May 2020 21:19:58 GMT
Connection: close
Content-Length: 96
Content-Type: application/json; charset=UTF-8
X-Weave-Timestamp: 1589404798.86
X-Timestamp: 1589404798
```
- "Unauthorized" expected since for sync the client needs to authenticate against a Mozilla account server to get an access token.

CedArctic commented:
@MichaIng testing ASAP, I'll report back my findings.

CedArctic commented:
@MichaIng I've tested and ran into the following:
`firefox-sync.service - Firefox Sync Server (DietPi)                                                                │ 
                                          │    Loaded: loaded (/etc/systemd/system/firefox-sync.service; disabled; vendor preset: enabled)                       │ 
                                          │    Active: failed (Result: exit-code) since Thu 2020-05-14 10:23:04 BST; 4min 12s ago                                │ 
                                          │      Docs: https://github.com/mozilla-services/syncserver                                                            │ 
                                          │   Process: 1067 ExecStart=/opt/firefox-sync/local/bin/gunicorn --paste                                               │ 
                                          │ /mnt/dietpi_userdata/firefox-sync/syncserver.ini (code=exited, status=1/FAILURE)                                     │ 
                                          │  Main PID: 1067 (code=exited, status=1/FAILURE)                                                                      │ 
                                          │                                                                                                                      │ 
                                          │ May 14 10:22:58 DietPi systemd[1]: Started Firefox Sync Server (DietPi).                                             │ 
                                          │ May 14 10:23:04 DietPi gunicorn[1067]: Error: [Errno 13] Permission denied:                                          │ 
                                          │ '/mnt/dietpi_userdata/firefox-sync/syncserver.ini'                                                                   │ 
                                          │ May 14 10:23:04 DietPi systemd[1]: firefox-sync.service: Main process exited, code=exited, status=1/FAILURE          │ 
                                          │ May 14 10:23:04 DietPi systemd[1]: firefox-sync.service: Failed with result 'exit-code'.  `

The issue seems to be:
`DietPi gunicorn[1067]: Error: [Errno 13] Permission denied: │ │ '/mnt/dietpi_userdata/firefox-sync/syncserver.ini'`

MichaIng commented:
@CedArctic 
Please try:
```
chmod a+x /mnt/dietpi_userdata
systemctl restart firefox-sync
```
AFAIK older default `/mnt/dietpi_userdata` modes were `770`. Now instead of adding all services to `dietpi` group to allow them accessing their subdir (but as well full write permissions to whole `/mnt/dietpi_userdata`) it is better to make it world-executable, so even that ffsync does not have read permissions, it can access it's own home dir where it has all permissions it needs.
________
This should do it: https://github.com/MichaIng/DietPi/commit/cf7351ad9423745bfde4fe352ff26859aa235f48

MichaIng committed:
v6.31

+ DietPi-Patch | Make userdata dir world-executable so service users don't need to be in dietpi group to access their data dir: https://github.com/MichaIng/DietPi/pull/3536#issuecomment-628515444

CedArctic commented:
@MichaIng tried the commands, however it didn't seem to fix the issue. I'll try reinstalling later today and see if this was fixed

MichaIng commented:
@CedArctic 
Hmm, can you paste the following:
```
ls -al /mnt/dietpi_userdata /mnt/dietpi_userdata/firefox-sync
```

CedArctic commented:
@MichaIng Sure, here you go:
`
dietpi@DietPi:~$ ls -al /mnt/dietpi_userdata /mnt/dietpi_userdata/firefox-sync
lrwxrwxrwx 1 root   root   24 Mar 24 14:01 /mnt/dietpi_userdata -> /mnt/hdd/dietpi_userdata

/mnt/dietpi_userdata/firefox-sync:
total 16
drwxr-xr-x 1 ffsync ffsync    0 May 14 10:03 .
drwxr-xr-x 1 dietpi dietpi 4096 May 14 10:03 ..
-rw-r--r-- 1 ffsync ffsync 8192 May 14 10:03 FF-Sync-DB.db
---------- 1 ffsync ffsync 2186 May 14 10:03 syncserver.ini
`

MichaIng commented:
@CedArctic 
Ah, somehow `syncserver.ini` has 000 permissions 😄, how did this happen. Do:
```
chmod 644 /mnt/dietpi_userdata/firefox-sync/syncserver.ini
systemctl restart firefox-sync
```
Luckily this broad my attention to `/mnt/dietpi_userdata` permissions which can break access as well.
_____
We do:
```
cp -a /opt/firefox-sync/syncserver.ini $ffsync_conf
```
So existing permissions are preserved. Can you paste the output of:
```
ls -al /opt/firefox-sync
```

CedArctic commented:
@MichaIng 

ls -al /opt/firefox-sync output:

dietpi@DietPi:~$ ls -al /opt/firefox-sync
total 116
drwxrwxr-x 7 root root  4096 May 14 10:03 .
drwxr-xr-x 6 root root  4096 May 14 09:53 ..
drwxrwxr-x 3 root root  4096 Mar 30 19:58 bin
drwxrwxr-x 2 root root  4096 Mar 30 19:58 .circleci
-rw-rw-r-- 1 root root   691 Mar 30 19:58 CODE_OF_CONDUCT.md
-rw-rw-r-- 1 root root  3340 Mar 30 19:58 CONTRIBUTING.md
-rw-rw-r-- 1 root root    24 Mar 30 19:58 dev-requirements.txt
-rwxrwxr-x 1 root root  1122 Mar 30 19:58 docker-entrypoint.sh
-rw-rw-r-- 1 root root   751 Mar 30 19:58 Dockerfile
-rw-rw-r-- 1 root root    71 Mar 30 19:58 .dockerignore
-rw-rw-r-- 1 root root    81 Mar 30 19:58 .gitignore
-rw-rw-r-- 1 root root 16726 Mar 30 19:58 LICENSE
drwxr-xr-x 7 root root  4096 May 14 10:03 local
-rw-rw-r-- 1 root root  1861 Mar 30 19:58 Makefile
-rw-rw-r-- 1 root root    76 Mar 30 19:58 MANIFEST.in
-rw-rw-r-- 1 root root     7 Mar 30 19:58 .python-version
-rw-rw-r-- 1 root root  6036 Mar 30 19:58 README.rst
-rw-rw-r-- 1 root root   329 Mar 30 19:58 requirements.txt
-rw-rw-r-- 1 root root   408 Mar 30 19:58 setup.py
drwxrwxr-x 2 root root  4096 Mar 30 19:58 syncserver
drwxr-xr-x 2 root root  4096 May 14 10:03 syncserver.egg-info
-rw-rw-r-- 1 root root  2150 Mar 30 19:58 syncserver.ini
-rw-rw-r-- 1 root root  1204 Mar 30 19:58 syncserver.wsgi
-rw-rw-r-- 1 root root   588 Mar 30 19:58 .travis.yml

MichaIng commented:
@CedArctic 
Very strange:
```
-rw-rw-r-- 1 root root 2150 Mar 30 19:58 syncserver.ini
```
No idea who it could became 000 permissions. Could you try:
```
cp -a /opt/firefox-sync/syncserver.ini /mnt/dietpi_userdata/firefox-sync/test
ls -l /mnt/dietpi_userdata/firefox-sync/test
```

CedArctic commented:
@MichaIng will do in a few minutes. By the way I've managed to get it syncing now, we need to set force_wsgi_environ = true to get things working.

CedArctic commented:
@MichaIng here's the output:

-rw------- 1 root root 2150 Mar 30 19:58 /mnt/dietpi_userdata/firefox-sync/test

MichaIng commented:
@CedArctic 
Very strange, cp fails to apply permissions, more strange that the result is different then the last time. As you moved userdata to an external drive, which filesystem does it have?
```
df -T
```
______________
> we need to set force_wsgi_environ = true to get things working.

Ah this matches the docs, when you setup a proxy: https://github.com/SynoCommunity/spksrc/wiki/Firefox-Sync-Server-1.5#firefox-sync-changes
Little background, make sense: https://en.wikipedia.org/wiki/Web_Server_Gateway_Interface
However if you access the sync server on port 5000 directly, the setting should not be required.

CedArctic commented:
@MichaIng on WSGI it seemed weird to me too, but it's the only way to get the android client syncing, PC's don't seem to have the issue. Don't know if it's anything on my side that I'm missing.

My drive is NTFS, df -T yields:

Filesystem     Type     1K-blocks      Used Available Use% Mounted on
/dev/root      ext4       7345680   4974076   2048060  71% /
devtmpfs       devtmpfs    494908         0    494908   0% /dev
tmpfs          tmpfs       499516       260    499256   1% /dev/shm
tmpfs          tmpfs       499516      6660    492856   2% /run
tmpfs          tmpfs         5120         0      5120   0% /run/lock
tmpfs          tmpfs       499516         0    499516   0% /sys/fs/cgroup
tmpfs          tmpfs        51200         0     51200   0% /var/log
tmpfs          tmpfs      1048576         0   1048576   0% /tmp
/dev/mmcblk0p1 vfat        258095     56618    201478  22% /boot
/dev/sdb1      fuseblk  488383484 114068100 374315384  24% /mnt/hdd

MichaIng commented:
@CedArctic 
My goodness, unluck within luck, or the other way round, but we found a very important (external) bug, as I was able to replicate the issue on NTFS:
```
2020-05-14 19:54:50 root@VM-Buster:~# l
total 24K
-rw------- 1 root root 4.1K May 10 12:36 .bash_history
drwx------ 3 root root 4.0K Jan  7 16:21 .config
drwxr-xr-x 3 root root 4.0K Feb 21 22:55 .local
drwxr-xr-x 2 root root 4.0K Jan  7 15:08 .ssh
-rw-r--r-- 1 root root  294 May 10 12:30 .wget-hsts
2020-05-14 19:54:56 root@VM-Buster:~# cp -a .bash_history /mnt/external/
'.bash_history' -> '/mnt/external/.bash_history'
2020-05-14 19:55:07 root@VM-Buster:~# l /mnt/external/
total 8.0K
-rw------- 1 root root 4.1K May 10 12:36 .bash_history
2020-05-14 19:55:10 root@VM-Buster:~# cp -a .local/ /mnt/external/
'.local/' -> '/mnt/external/.local'
'.local/share' -> '/mnt/external/.local/share'
'.local/share/nano' -> '/mnt/external/.local/share/nano'
'.local/share/nano/search_history' -> '/mnt/external/.local/share/nano/search_history'
2020-05-14 19:55:21 root@VM-Buster:~# l /mnt/external/
total 8.0K
-rw------- 1 root root 4.1K May 10 12:36 .bash_history
drwx------ 1 root root  144 Feb 21 22:55 .local
2020-05-14 19:55:25 root@VM-Buster:~# chmod 755 /mnt/external/.local
2020-05-14 19:56:13 root@VM-Buster:~# l /mnt/external/
total 8.0K
-rw------- 1 root root 4.1K May 10 12:36 .bash_history
drwxr-xr-x 1 root root  144 Feb 21 22:55 .local
```
/mnt/external is a NTFS mount. As can be seen the first file `.bash_history` kept its permissions, the dir `.local` lost its group and others mode. Seems like the fuse/ntfs-3g method to emulate UNIX permissions is somehow unreliable or has some limitations in cases. Direkt chmod/chown works fine, but the cp -a/-p option not. This is very bad as this expected to simply work and many things completely rely on this fundamental command option.
_________
The consequence is very bad... We cannot set all permissions or all files again just to cover NTFS, some downloads contain hundreds of files, with individual permissions, changing regularly with new versions, impossible to not rely on cp -a/-p... I'll test this on most current version (Bullseye) and in case report to Debian, in case to upstream devs. Let's see if this is somehow expected or a known issue. At least searching through the web does not give any hints...

CedArctic commented:
@MichaIng this is unfortunate, however quite interesting. Should we bring this to the attention of fuse? In the mean time for our own purposes we can manually apply permissions.

MichaIng commented:
@CedArctic 
See my edit above. I don't want to manually apply permissions in one case, while it is clearly impossible to do it in all cases, so this causes inconsistencies. If there is no solution to get cp -a/-p working reliable, I would rather unallow to move dietpi_userdata onto an NTFS drive 😢. I mean it is anyway slow and CPU intense and not recommended, although I know that it is very nice that it can be simply plugged into a Windows system... However lets see what the maintainers/devs come up with.
__________
Further details about WSGI:
```
# Set this to "true" to work around a mismatch between public_url and
# the application URL as seen by python, which can happen in certain reverse-
# proxy hosting setups.  It will overwrite the WSGI environ dict with the
# details from public_url.  This could have security implications if e.g.
# you tell the app that it's on HTTPS but it's really on HTTP, so it should
# only be used as a last resort and after careful checking of server config.
force_wsgi_environ = false
```
Does your `public_url` actually contain the public IP/domain or probably the local network IP? I think we'll leave it disabled by default and instead add it to our online docs to try if clients fail to connect.

CedArctic commented:
@MichaIng public_url is set to the Pi's IP in the local network. Let's leave it as is for now with the note in the docs you mentioned.

It's a bit of a bummer that NTFS can't be used like this.

MichaIng commented:
@CedArctic 
> public_url is set to the Pi's IP in the local network. Let's leave it as is for now with the note in the docs you mentioned.

Yes, I agree, default on install + docs is fine like this. I was just thinking if your config probably contain the local IP while you access from Android phone from outside the local network via public IP/domain. In this case, AFAIK, the setting needs to be changed accordingly, not sure if related to WSGI though 😉.

> It's a bit of a bummer that NTFS can't be used like this.

Indeed. However since chmod/chown works, things can be fixed manually, in case of ffsync simple and safe, because it's only a single file. Let's hope that there is a solution.

MichaIng commented:
Okay I'll merge it now. Info has been added to online docs, NTFS topic is not related, I'll open a separate issue and bug report about this.
